src/main/java/org/truffleruby/extra/ConcurrentHashMapNodes.java

```diff
@@ -0,0 +1,300 @@
+            final RubyConcurrentHashMap instance = new RubyConcurrentHashMap(rubyClass, shape, null);
+            AllocationTracing.trace(instance, this);
+            return instance;
+        }
+    }
+
+    @CoreMethod(names = "initialize", optional = 1)
+    public abstract static class InitializeNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        protected RubyConcurrentHashMap initializeNoValue(RubyConcurrentHashMap self, NotProvided options) {
+            return self;
+        }
+
+        @Specialization
+        protected RubyConcurrentHashMap initializeNil(RubyConcurrentHashMap self, Nil options) {
+            return self;
+        }
+
+        @Specialization
+        protected RubyConcurrentHashMap initializeWithOptions(RubyConcurrentHashMap self, RubyHash options) {
+            return self;
+        }
+
+        @Specialization(guards = { "!isNil(options)", "wasProvided(options)" })
+        protected RubyConcurrentHashMap initializeWithOptions(RubyConcurrentHashMap self, RubyBasicObject options) {
+            return self;
+        }
+    }
+
+    @CoreMethod(names = "initialize_copy", required = 1)
+    public abstract static class InitializeCopyNode extends CoreMethodArrayArgumentsNode {
+        @Specialization
+        protected RubyConcurrentHashMap initializeCopy(RubyConcurrentHashMap self, RubyConcurrentHashMap other) {
+            self.concurrentHash.putAll(self.concurrentHash);
+            return self;
+        }
+    }
+
+    @CoreMethod(names = "[]", required = 1)
+    public abstract static class GetIndexNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object getIndex(RubyConcurrentHashMap self, Object key) {
+            final Object value = self.concurrentHash.get(new RubyConcurrentHashMap.Key(key));
+            if (value == null) {
+                return nil;
+            } else {
+                return value;
+            }
+        }
+    }
+
+    @CoreMethod(names = "[]=", required = 2)
+    public abstract static class SetIndexNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object setIndex(RubyConcurrentHashMap self, Object key, Object value) {
+            self.concurrentHash.put(new RubyConcurrentHashMap.Key(key), value);
+            return value;
+        }
+    }
+
+    @CoreMethod(names = "compute_if_absent", required = 1, needsBlock = true)
+    public abstract static class ComputeIfAbsentNode extends YieldingCoreMethodNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object computeIfAbsent(RubyConcurrentHashMap self, Object key, RubyProc block) {
+            Object returnValue = self.concurrentHash.computeIfAbsent(new RubyConcurrentHashMap.Key(key), k -> {
+                return callBlock(block);
+            });
+
+            if (returnValue == null) {
```

Comment:
The return value should never be `null` here (it's either the value in the map, or the result of the blocks and both can't be `null`), so we don't need this check. We can have an `assert returnValue != null` for clarity.